www/js/config/dynamic_config.js

```diff
@@ -164,23 +164,93 @@ angular.module('emission.config.dynamic', ['emission.plugin.logger'])
         }
     }
 
+    /*
+     * We want to support both old style and new style tokens.
+     * Theoretically, we don't need anything from this except the study
+     * name, but we should re-validate the token for extra robustness.
+     * The control flow here is a bit tricky, though.
+     * - we need to first get the study name
+     * - then we need to retrieve the study config
+     * - then we need to re-validate the token against the study config,
+     * and the subgroups in the study config, in particular.
+     *
+     * So let's support two separate functions here - extractStudyName and extractSubgroup
+     */
+    dc.extractStudyName = function(token) {
+        const tokenParts = token.split("_");
+        if (tokenParts.length < 3) {
+          // all tokens must have at least nrelop_[study name]_...
+          throw new Error("token "+token+" does not have at least two '_' characters");
+        }
+        if (tokenParts[0] != "nrelop") {
+          throw new Error("token "+token+" does not start with 'nrelop', please re-check");
+        }
+        return tokenParts[1];
+    }
+
+    dc.extractSubgroup = function(token, config) {
+        if (config.opcode) {
+            // new style study, expects token with sub-group
+            const tokenParts = token.split("_");
+            if (tokenParts.length <= 3) { // no subpart defined
+                throw new Error("Invalid opcode format, expected 'nrelop_study_subgroup_[random string]");
+            }
+            if (config.opcode.subgroups) {
+                if (config.opcode.subgroups.indexOf(tokenParts[2]) == -1) {
+                // subpart not in config list
+                    throw new Error("Invalid opcode, subgroup '"+tokenParts[2]+"' not found in list '"+config.opcode.subgroups+"'");
+                } else {
+                    console.log("subgroup "+tokenParts[2]+" found in list");
+                    return tokenParts[2];
+                }
+            } else {
+                if (tokenParts[2] != "default") {
```

Comment:
So all OPCodes from now on **must** include a subgroup?
But if subgroups aren't being used, it must be "default"?

Reply:
All new style tokens, yes.